components/salsa-2022/src/storage.rs

```diff
@@ -135,19 +139,16 @@ where
         loop {
             self.runtime.set_cancellation_flag();
 
+            // jars need to be protected by a lock to avoid deadlocks.
+            let mut guard = self.shared.noti_lock.lock();
```

Comment:
Can you spell this out for me, I want to be sure I understand the deadlock...

Reply:
> Can you spell this out for me, I want to be sure I understand the deadlock...

Well, suppose there are two threads A and B, A is holding the original database and B is holding a snapshot, then B wants to make some changes and calls the `cancel_other_workers` function. When the code runs to line 139 the B thread is not exited yet, so `Arc::get_mut` returns `None` and continues. Just before it reaches line 151, thread B exits and the `self.shared.cvar.notify_all();` on line 182 was called, but it did nothing as there's no thread waiting for a signal right now. Then A thread reaches line 151, waiting for a signal that would never be triggered.

https://github.com/salsa-rs/salsa/blob/4151b09d3c37e50f1a01894fcdfa776bc2909132/components/salsa-2022/src/storage.rs#L129-L153

https://github.com/salsa-rs/salsa/blob/4151b09d3c37e50f1a01894fcdfa776bc2909132/components/salsa-2022/src/storage.rs#L174-L184

Reply:
I see. Duh. Makes sense!